### PR TITLE
Adds Cyclical Learning Rates

### PR DESCRIPTION
**Pull request by thomasjpfan**
Adds feature requested in https://github.com/pytorch/pytorch/issues/1909. Mimics the parameters from https://github.com/bckenstler/CLR. Since Cyclical Learning Rate (CLR) requires updating the learning rate after every batch, I added `batch_step` which should be called after feeding a batch through a neutral network.

### CODE REVIEW

**Comment by Jiaming-Liu** [comment]
The usage of lambda would make the instance of CycleLR not pickle-able. Also, it would be nice if this can be written as a subclass of _LRScheduler.

**Comment by thomasjpfan** [comment]
`_LRScheduler` interface seems to imply that `step` is called at the end of each epoch. The cyclical learning rate policy updates the learning rate after every batch. If `CyclicLR` were to subclass `_LRSchleduler` it would have to raise an exception for calling `step` and ask the user to call `batch_step` instead. Basically it comes down to a design decision.

**Comment by Jiaming-Liu** [comment]
You are right -- some scheduling methods require a different calling frequency. The main discrepancy would be the timing to call `.step()` and the name of some methods/attributes/parameters. It might be helpful if there is another abstract class for batch-wise scheduler. @apaszke Any suggestion?

**Comment by thomasjpfan** [comment]
I think the only thing stopping this PR is the added complexity of the `batch_step` function to `CyclicLR`. All the other schedulers are updated after every epoch, while `CyclicLR` is updated every batch.

**Comment by Aphaniteja** [comment]
I hope this gets added. Recent papers(https://github.com/timgaripov/swa) are building upon this concept.

**Comment by thomasjpfan** [comment]
https://github.com/timgaripov/swa could be implemented in this `_LRScheduler` framework. The `step` function would have to return the weight of the moving average, and the user would have to update the moving average of the model weights themselves.

**Comment by sean-adler** [comment]
> You are right -- some scheduling methods require a different calling frequency. The main discrepancy would be the timing for `.step()` and the name of some methods/attributes/parameters. It might be helpful if there is another abstract class for batch-wise scheduler. @apaszke Any suggestion?

Sorry to ask a newbie question - is it necessary to have a different API for updating the learning rate per-batch rather than per-epoch?

If `CyclicLR` were rewritten to subclass `_LRScheduler` and implement `step()` instead of `batch_step()`, it kind of seems like doing this would be fine:

```
for epoch in range(num_epochs):
    for batch in data_loader:
        optimizer.zero_grad()
        output = model(input)
        loss = criterion(output, target)
        loss.backward()

        optimizer.step()
        scheduler.step()  #  <- apply CLR per-batch
```

Because `step_size` is an argument to `CyclicLR` in this changeset, making cycles follow a schedule similar to those in [the paper](https://arxiv.org/pdf/1506.01186.pdf) by e.g. setting `step_size` to a multiple of `len(data_loader)` seems like it'd make the above work (and be reasonably straightforward).

(One downside of doing this is that `CyclicLR` would have a misleading `last_epoch` attribute, but I have no idea how big a deal that is.)

I have very little pytorch-specific context here, so any feedback would be hugely appreciated.

**Comment by EtienneDesticourt** [comment]
Very useful stuff, can't wait for this PR to be merged!  

I've been using it to train my models and I think the ``step_size`` parameter is confusing though. From its name you would expect it to be the amount of change in the learning rate at each step i.e: ``step_size = (max_lr - min_lr) / num_steps_per_cycle``. Whereas in this implementation ``step_size`` is actually the number of steps per half cycle (I also think the "number of training iterations per half cycle" part of the description is needlessly cryptic, speaking in terms of steps makes it much clearer, the mental connection between training iterations and steps happens when you call batch_step either way.)

So to sum up I propose something like this instead:  
`
num_steps(int): Number of steps per half cycle. Authors suggest setting num_steps
                         2-8 x training steps in epoch. Default: 2000
`

**Comment by sai-prasanna** [comment]
Is there a way to find the optimal learning rate as mentioned in that paper in this commit? If not,  If I wish to add that, which part of torch would be best suited ?

**Comment by Randl** [comment]
What's up with this PR?

**Comment by Randl** [comment]
@thomasjpfan @apaszke What if we add `batch_step` to `_LRScheduler`? It would add some complexity, but at least it will be uniform along different schedulers. The code of form
```python3
for epoch in epochs:
    scheduler.step()
    ...
    for batch in batches:
        scheduler.batch_step()
        ...
```
would work for all schedulers then.

**Comment by apaszke** [comment]
But that's so verbose... If we need a per-batch callback we might want to write a wrapper around an iterator that returns the batches (e.g. `DataLoader`)

**Comment by thomasjpfan** [comment]
@sethah Good point. I agree that it would be better to just let the user decide when to call `step`. I will update this PR with that in mind.

**Comment by slowbull** [comment]
Have anyone tried it on training cifar or imagenet? any code available?

**Comment by suruoxi** [comment]
@slowbull  You can check https://github.com/Randl/MobileNetV2-pytorch

**Comment by Randl** [comment]
ping reviewers

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -4,6 +4,8 @@
 from torch._six import inf
 from bisect import bisect_right
 from functools import partial
+
+import numpy as np
```
Comment (soumith) [inline torch/optim/lr_scheduler.py]: numpy is an optional dependency into PyTorch.
Please remove the dependence of np in this file.

File: test/test_optim.py [hunk 2]
```diff
@@ -640,15 +642,125 @@ def test_reduce_lr_on_plateau8(self):
...
+                             step_size_up=2, step_size_down=6,
+                             mode='exp_range', gamma=gamma)
+        self._test_cycle_lr(scheduler, targets, len(target))
+
     def test_lambda_lr(self):
         epochs = 10
         self.opt.param_groups[0]['lr'] = 0.05
         self.opt.param_groups[1]['lr'] = 0.4
-        targets = [[0.05 * (0.9 ** x) for x in range(epochs)], [0.4 * (0.8 ** x) for x in range(epochs)]]
+        targets = [[0.05 * (0.9 ** x) for x in range(epochs)], [0.4 * (0.8 ** x)
```
Comment (soumith) [inline test/test_optim.py]: please only change lines relevant to your PR.
We have line limits of 120 characters, so your changes here and above in this file are not needed, and actively introduce noise.

File: torch/optim/lr_scheduler.py [hunk 3]
```diff
@@ -413,3 +415,166 @@ def state_dict(self):
...
+            Values correspond to policies detailed above.
+            If scale_fn is not None, this argument is ignored.
+            Default: 'triangular'
+        gamma (float): Constant in 'exp_range' scaling function:
+            gamma**(cycle iterations)
+            Default: 1.0
+        scale_fn (function): Custom scaling policy defined by a single
+            argument lambda function, where
+            0 <= scale_fn(x) <= 1 for all x >= 0.
+            mode paramater is ignored
```
Comment (soumith) [inline torch/optim/lr_scheduler.py]: *parameter

File: torch/optim/lr_scheduler.py [hunk 4]
```diff
@@ -413,3 +415,166 @@ def state_dict(self):
...
+            lower boundary in the cycle for eachparam groups.
+            Default: 0.001
+        max_lr (float or list): Upper boundaries in the cycle for
+            each parameter group. Functionally,
+            it defines the cycle amplitude (max_lr - base_lr).
+            The lr at any cycle is the sum of base_lr
+            and some scaling of the amplitude; therefore
+            max_lr may not actually be reached depending on
+            scaling function. Default: 0.006
+        step_size_up (int): Number of training iterations in the
```
Comment (soumith) [inline torch/optim/lr_scheduler.py]: this has a default value below, but doc is missing

File: torch/optim/lr_scheduler.py [hunk 5]
```diff
@@ -413,3 +415,166 @@ def state_dict(self):
...
+            increasing half of a cycle.
+        step_size_down (int): Number of training iterations in the
```
Comment (soumith) [inline torch/optim/lr_scheduler.py]: this has a default value below, but doc is missing

**Review by soumith — CHANGES_REQUESTED** [review]

**Comment by klicperajo** [comment]
This only implements a cyclical LR change. However, both the Leslie Smith paper that introduced the 1cycle schedule (https://arxiv.org/pdf/1803.09820.pdf) and the fast.ai library (https://docs.fast.ai/callbacks.one_cycle.html#Training-with-the-1cycle-policy) recommend to decrease the momentum as the learning rate increases, which helps with convergence. Would this be possible to add?

**Comment by thomasjpfan** [comment]
The 1cycle policy can be configured by setting the `step_size_up` to half the total number of training iterations. Yes, it is possible to add the momentum piece to this PR.

**Comment by soumith** [comment]
finished by https://github.com/pytorch/pytorch/pull/18001 which is landing
